plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2600,12 +2757,196 @@ void producer_plugin::log_failed_transaction(const transaction_id_type& trx_id,
             ("entire_trx", packed_trx_ptr ? my->chain_plug->get_log_trx(packed_trx_ptr->get_transaction()) : fc::variant{trx_id}));
 }
 
-void producer_plugin::register_metrics_listener(metrics_listener listener) {
-   my->_metrics.register_listener(listener);
+// Called from app thread
+void producer_plugin_impl::switch_to_write_window() {
+   // this method can be called from multiple places. it is possible
+   // we are already in write window.
+   if ( app().executor().is_write_window() ) {
+      return;
+   }
+
+   EOS_ASSERT(_ro_num_active_trx_exec_tasks.load() == 0 && _ro_trx_exec_tasks_fut.empty(), producer_exception, "no read-only tasks should be running before switching to write window");
+   _ro_read_window_timer.cancel();
+   _ro_write_window_timer.cancel();
+
+   start_write_window();
+}
+
+// Called from app thread
+void producer_plugin_impl::start_write_window() {
+   chain::controller& chain = chain_plug->chain();
+
+   app().executor().set_to_write_window();
+   chain.unset_db_read_only_mode();
+   _idle_trx_time = fc::time_point::now();
+
+   auto expire_time = boost::posix_time::microseconds(_ro_write_window_time_us.count());
+   _ro_write_window_timer.expires_from_now( expire_time );
```

Comment:
Consider using only one timer since only one window should be active at any given time.